File: crates/pdl-core/src/world_a.rs
//! World A expression-tree authoring — desugar to classic declare-and-mount AST.
//! See `docs/PROPOSAL_WORLD_A_EXPRESSION_TREES.md`.
//!
//! Rust port of `src/worldA.ts`.

use indexmap::IndexMap;

use crate::ast::{ChildEntry, FrameBodyItem, ValueExpr};

pub const FRAME_CTOR_NAMES: &[&str] = &["Text", "Layout", "Icon", "Media"];

/// Reserved component names — collide with World A frame ctors.
pub const RESERVED_FRAME_CTOR_COMPONENT_NAMES: &[&str] = FRAME_CTOR_NAMES;

pub fn is_frame_ctor_name(name: &str) -> bool {
    FRAME_CTOR_NAMES.contains(&name)
}

/// Map World A ctor spelling → classic frame kind keyword.
pub fn frame_ctor_to_kind(name: &str) -> Option<&'static str> {
    match name {
        "Text" => Some("text"),
        "Layout" => Some("layout"),
        "Icon" => Some("icon"),
        "Media" => Some("media"),
        _ => None,
    }
}

/// Build a classic frame-let body from World A ctor kwargs (`children:` handled separately).
pub fn frame_ctor_kwargs_to_body(
    args: &IndexMap<String, ValueExpr>,
    child_entries: Option<Vec<ChildEntry>>,
) -> Vec<FrameBodyItem> {
    let mut body: Vec<FrameBodyItem> = Vec::new();
    for (name, value) in args {
        if name == "children" {
            continue;
        }
        body.push(FrameBodyItem::Prop {
            name: name.clone(),
            value: value.clone(),
        });
    }
    if let Some(entries) = child_entries {
        if !entries.is_empty() {
            body.push(FrameBodyItem::Children {
                target: crate::ast::ChildrenTarget::Root,
                entries,
            });
        }
    }
    body
}

/// Lower World A `frameCtor` child entries into synthetic `let` + `frameRef` mounts.
/// Nested ctors hoist before their parent so E019 order stays valid.
pub fn lower_world_a_body(body: Vec<FrameBodyItem>) -> Vec<FrameBodyItem> {
    let mut auto = 0usize;
    lower_items(body, &mut auto)
}

fn next_id(kind: &str, auto: &mut usize) -> String {
    let id = format!("__auto_{kind}_{auto}");
    *auto += 1;
    id
}

fn lower_entries(
    entries: Vec<ChildEntry>,
    hoist: &mut Vec<FrameBodyItem>,
    auto: &mut usize,
) -> Vec<ChildEntry> {
    entries
        .into_iter()
        .map(|entry| match entry {
            ChildEntry::FrameCtor {
                frame_kind,
                props,
                child_entries,
                opacity,
            } => {
                let nested_children = child_entries.map(|ce| lower_entries(ce, hoist, auto));
                let id = next_id(&frame_kind, auto);
                let let_body = frame_ctor_kwargs_to_body(&props, nested_children);
                hoist.push(FrameBodyItem::Let {
                    id: id.clone(),
                    frame_kind,
                    body: lower_items(let_body, auto),
                });
                ChildEntry::FrameRef { id, opacity }
            }
            other => other,
        })
        .collect()
}

fn lower_items(items: Vec<FrameBodyItem>, auto: &mut usize) -> Vec<FrameBodyItem> {
    let mut out: Vec<FrameBodyItem> = Vec::new();
    for item in items {
        match item {
            FrameBodyItem::Children { target, entries } => {
                let mut hoist: Vec<FrameBodyItem> = Vec::new();
                let entries = lower_entries(entries, &mut hoist, auto);
                out.extend(hoist);
                out.push(FrameBodyItem::Children { target, entries });
            }
            FrameBodyItem::Let {
                id,
                frame_kind,
                body,
            } => {
                out.push(FrameBodyItem::Let {
                    id,
                    frame_kind,
                    body: lower_items(body, auto),
                });
            }
            FrameBodyItem::If { chain } => {
                let mut chain = chain;
                chain.branches = chain
                    .branches
                    .into_iter()
                    .map(|mut b| {
                        b.body = lower_items(b.body, auto);
                        b
                    })
                    .collect();
                if let Some(else_body) = chain.else_body {
                    chain.else_body = Some(lower_items(else_body, auto));
                }
                out.push(FrameBodyItem::If { chain });
            }
            other => out.push(other),
        }
    }
    out
}

File: src/worldA.ts
/**
 * World A expression-tree authoring — desugar to classic declare-and-mount AST.
 * See docs/PROPOSAL_WORLD_A_EXPRESSION_TREES.md.
 */

import type { ChildEntry, FrameBodyItem, ValueExpr } from "./ast.js";

export const FRAME_CTOR_NAMES = new Set(["Text", "Layout", "Icon", "Media", "Presenter"]);
export type FrameCtorName = "Text" | "Layout" | "Icon" | "Media" | "Presenter";

export const FRAME_CTOR_TO_KIND: Record<
  FrameCtorName,
  "layout" | "text" | "icon" | "media" | "presenter"
> = {
  Text: "text",
  Layout: "layout",
  Icon: "icon",
  Media: "media",
  Presenter: "presenter",
};

export function isFrameCtorName(name: string): name is FrameCtorName {
  return FRAME_CTOR_NAMES.has(name as FrameCtorName);
}

/** Reserved component names — collide with World A frame ctors. */
export const RESERVED_FRAME_CTOR_COMPONENT_NAMES = FRAME_CTOR_NAMES;

/** Build a classic frame-let body from World A ctor kwargs (`children:` handled separately). */
export function frameCtorKwargsToBody(
  args: Record<string, ValueExpr>,
  childEntries?: ChildEntry[],
): FrameBodyItem[] {
  const body: FrameBodyItem[] = [];
  for (const [name, value] of Object.entries(args)) {
    if (name === "children") continue;
    body.push({ kind: "prop", name, value });
  }
  if (childEntries && childEntries.length > 0) {
    body.push({ kind: "children", target: "root", entries: childEntries });
  }
  return body;
}

/**
 * Lower World A `frameCtor` child entries into synthetic `let` + `frameRef` mounts.
 * Nested ctors hoist before their parent so E019 order stays valid.
 */
export function lowerWorldABody(body: FrameBodyItem[]): FrameBodyItem[] {
  let auto = 0;
  const nextId = (kind: string) => `__auto_${kind}_${auto++}`;

  function lowerEntries(entries: ChildEntry[], hoist: FrameBodyItem[]): ChildEntry[] {
    return entries.map((entry) => {
      if (entry.kind !== "frameCtor") return entry;
      const nestedChildren = entry.childEntries
        ? lowerEntries(entry.childEntries, hoist)
        : undefined;
      const id = nextId(entry.frameKind);
      const letBody = frameCtorKwargsToBody(entry.props, nestedChildren);
      hoist.push({
        kind: "let",
        id,
        frameKind: entry.frameKind,
        body: lowerItems(letBody),
      });
      return {
        kind: "frameRef" as const,
        id,
        ...(entry.opacity ? { opacity: entry.opacity } : {}),
      };
    });
  }

  function lowerItems(items: FrameBodyItem[]): FrameBodyItem[] {
    const out: FrameBodyItem[] = [];
    for (const item of items) {
      if (item.kind === "children") {
        const hoist: FrameBodyItem[] = [];
        const entries = lowerEntries(item.entries, hoist);
        out.push(...hoist);
        out.push({ ...item, entries });
        continue;
      }
      if (item.kind === "let") {
        out.push({ ...item, body: lowerItems(item.body) });
        continue;
      }
      if (item.kind === "if") {
        out.push({
          kind: "if",
          chain: {
            branches: item.chain.branches.map((b) => ({
              ...b,
              body: lowerItems(b.body),
            })),
            ...(item.chain.elseBody
              ? { elseBody: lowerItems(item.chain.elseBody) }
              : {}),
          },
        });
        continue;
      }
      out.push(item);
    }
    return out;
  }

  return lowerItems(body);
}
